### PR TITLE
Exteranl zonemap

### PR DESCRIPTION
**Pull request by jianwan0214**
## What type of PR is this?

- [ ] API-change
- [ ] BUG
- [ ] Improvement
- [ ] Documentation
- [x] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue #7264 #6947 #6798 #6680 #6678

## What this PR does / why we need it:

### CODE REVIEW

**Review by daviszhen — APPROVED** [review]

File: pkg/vm/engine/disttae/util.go [hunk 1]
```diff
@@ -66,6 +66,10 @@ func getColumnMapByExpr(expr *plan.Expr, tableDef *plan.TableDef, columnMap *map
 	}
 }
 
+func GetColumnsByExpr(expr *plan.Expr, tableDef *plan.TableDef) (map[int]int, []int, int) {
```
Comment (nnsgmsone) [inline pkg/vm/engine/disttae/util.go:69]: Do not expose these functions. If you really need to expose. Find a special directory for it.

**Review by nnsgmsone — CHANGES_REQUESTED** [review]

File: pkg/sql/colexec/external/external.go [hunk 2]
```diff
@@ -650,6 +673,217 @@ func ScanFileData(param *ExternalParam, proc *process.Process) (*batch.Batch, er
 	return bat, nil
 }
 
+func getFileService(param *ExternalParam, dir string) (fileservice.FileService, error) {
```
Comment (reusee) [inline pkg/sql/colexec/external/external.go:676]: file service instance should always come from proc.FS

File: pkg/sql/colexec/external/external.go [hunk 3]
```diff
@@ -650,6 +673,217 @@ func ScanFileData(param *ExternalParam, proc *process.Process) (*batch.Batch, er
...
+	c := fileservice.Config{
+		Name:    defines.LocalFileServiceName,
+		Backend: "DISK",
+		DataDir: dir,
+	}
+	if param.extern.ScanType == tree.S3 {
+		c.Name = defines.S3FileServiceName
+		c.Backend = "S3"
```
Comment (reusee) [inline pkg/sql/colexec/external/external.go:684]: This is wrong. The backend of 'S3' should be configured in the config file. The initialization of proc.FS already handles these details. Do use proc.FS instead.

File: pkg/sql/colexec/external/external.go [hunk 4]
```diff
@@ -650,6 +673,217 @@ func ScanFileData(param *ExternalParam, proc *process.Process) (*batch.Batch, er
...
+		return getBatchFromZonemapFile(param, proc, objectReader)
+	} else {
+		return getBatchFromZonemapFile(param, proc, objectReader)
+	}
+}
+
+func ScanZonemapFile(param *ExternalParam, proc *process.Process) (*batch.Batch, error) {
+	if param.Filter.objectReader == nil {
+		dir, file := filepath.Split(param.extern.Filepath)
+		service, err := getFileService(param, dir)
```
Comment (reusee) [inline pkg/sql/colexec/external/external.go:834]: Should call fileservice.GetForETL to obtain a file service instance.

**Review by reusee — CHANGES_REQUESTED** [review]

File: pkg/fileservice/local_fs.go [hunk 5]
```diff
@@ -111,6 +111,10 @@ func NewLocalFS(
 	return fs, nil
 }
 
+func (l *LocalFS) SetRootpath(path string) {
```
Comment (reusee) [inline pkg/fileservice/local_fs.go:114]: I don't understand why this is required.
After initialization, a LocalFS should never change its root path.

File: pkg/sql/colexec/external/external.go [hunk 6]
```diff
@@ -667,22 +677,31 @@ func ScanCsvFile(param *ExternalParam, proc *process.Process) (*batch.Batch, err
...
-		c.Name = defines.S3FileServiceName
-		c.Backend = "S3"
-	} else if strings.HasPrefix(param.extern.Filepath, "etl:") {
-		c.Name = defines.ETLFileServiceName
-		c.Backend = "DISK-ETL"
+	if path.Service == "" {
+		path.Service = service.Name()
+	}
+	switch path.Service {
+	case "LOCAL":
```
Comment (reusee) [inline pkg/sql/colexec/external/external.go:689]: The LOCAL file service is global and its root path is configured in the config file. You should never change it.
Just use GetForETL. If GetForETL does not return what you want, fix it.

**Review by reusee — CHANGES_REQUESTED** [review]

**Review by ouyuanning — APPROVED** [review]

**Review by reusee — APPROVED** [review]

**Review by nnsgmsone — APPROVED** [review]

**Pull request merged**